django/core/handlers/asgi.py

```diff
@@ -174,65 +174,41 @@ async def handle(self, scope, receive, send):
-        request, error_response = self.create_request(scope, body_file)
-        if request is None:
-            body_file.close()
-            await self.send_response(error_response, send)
-            await sync_to_async(error_response.close)()
-            return
 
-        async def process_request(request, send):
-            response = await self.run_get_response(request)
-            try:
-                await self.send_response(response, send)
-            except asyncio.CancelledError:
-                # Client disconnected during send_response (ignore exception).
+        with closing(body_file):
+            # Request is complete and can be served.
+            set_script_prefix(get_script_prefix(scope))
+            await signals.request_started.asend(sender=self.__class__, scope=scope)
+            # Get the request and check for basic issues.
+            request, error_response = self.create_request(scope, body_file)
+            if request is None:
+                body_file.close()
+                await self.send_response(error_response, send)
+                await sync_to_async(error_response.close)()
+                return
+
+            class RequestProcessed(Exception):
                 pass
 
-            return response
-
-        # Try to catch a disconnect while getting response.
-        tasks = [
-            # Check the status of these tasks and (optionally) terminate them
-            # in this order. The listen_for_disconnect() task goes first
-            # because it should not raise unexpected errors that would prevent
-            # us from cancelling process_request().
-            asyncio.create_task(self.listen_for_disconnect(receive)),
-            asyncio.create_task(process_request(request, send)),
-        ]
-        await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
-        # Now wait on both tasks (they may have both finished by now).
-        for task in tasks:
-            if task.done():
-                try:
-                    task.result()
-                except RequestAborted:
-                    # Ignore client disconnects.
-                    pass
-                except AssertionError:
-                    body_file.close()
-                    raise
-            else:
-                # Allow views to handle cancellation.
-                task.cancel()
+            response = None
+            try:
                 try:
-                    await task
-                except asyncio.CancelledError:
-                    # Task re-raised the CancelledError as expected.
+                    async with asyncio.TaskGroup() as tg:
+                        tg.create_task(self.listen_for_disconnect(receive))
+                        response = await self.run_get_response(request)
+                        await self.send_response(response, send)
+                        raise RequestProcessed
+                except* (RequestProcessed, RequestAborted):
```

Comment:
should we add CancelledError here?

Reply:
Oh I see yes this PR needs reverting and this try/finally applying

Reply:
Yes, so we do want to fire request finished here. The client disconnected rather than completing the request, but we still need to clean up. 

I need to look at the tests here. The request finished signal not firing should trigger a failure. 🤔

Reply:
I don't think we should revert the current MR (it doesn't break the existing codebase), but we need to find a proper approach for handling the cleanup. We can move this discussion elsewhere.

Reply:
OK, so the signals.request_finished.asend is **still hit** on the tests checking cancellation. 

e.g. see `asgi.tests.ASGITest.test_asyncio_cancel_error`.

@Arfey If you could add a test case demonstrating a change in behaviour then we can certainly tweak.

Reply:
Thanks.

The `asgi.tests.ASGITest.test_asyncio_cancel_error` covers two cases:
- The request cycle completes normally when no disconnect is sent.
- The request cycle is interrupted when a disconnect occurs before the view responds.

So, it tests cancellation caused by a disconnection event. However, we discovered a potential cleanup issue if something higher up cancels `handle_request`.

>  demonstrating a change in behaviour then we can certainly tweak.

The behaviour remains the same. We already had the same potential issue in the previous version of the code.

> If you could add a test case

Sure

Reply:
Great. I'll look at that. Thanks. 

Not sure how such a case would come up currently. We can likely discuss elsewhere as you say.